moving Workspace tests to new framework

jmarolf opened the pull request.

jmarolf commented:
test windows_debug_vs-integration_prtest please

jmarolf commented:
test windows_release_vs-integration_prtest please

jmarolf commented:
test microbuild_prtest please

jmarolf commented:
@tmeschter @dotnet/roslyn-ide review please

jasonmalinowski commented on src/VisualStudio/IntegrationTest/IntegrationTests/AbstractEditorTest.cs:
```diff
@@ -138,10 +159,52 @@ protected void SendKeys(params object[] keys)
-            => VisualStudioWorkspaceOutOfProc.SetUseSuggestionMode(false);
+        {
+            VisualStudioWorkspaceOutOfProc.SetUseSuggestionMode(false);
+            WaitForAsyncOperations(FeatureAttribute.Workspace);
```
Why?

jasonmalinowski commented on src/VisualStudio/IntegrationTest/IntegrationTests/CSharp/CSharpQuickInfo.cs:
```diff
@@ -29,9 +29,8 @@ static void Main(string$$[] args)
-            Assert.Equal(
-                "class\u200e System\u200e.String\r\nRepresents text as a sequence of UTF-16 code units.To browse the .NET Framework source code for this type, see the Reference Source.",
-                Editor.GetQuickInfo());
+            VerifyQuickInfo(
```
I would have preferred it as is...

tmeschter replied:
Right. These one-line helper methods (like `VerifyQuickInfo` aren't making the tests any shorter, and they increase the complexity of the type...

jasonmalinowski commented on src/VisualStudio/IntegrationTest/IntegrationTests/Workspace/Workspaces.cs:
```diff
@@ -0,0 +1,71 @@
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.Workspace)]
+        public void AddDocuments()
```
What is this test actually asserting?

jasonmalinowski commented on src/VisualStudio/IntegrationTest/IntegrationTests/Workspace/Workspaces.cs:
```diff
@@ -0,0 +1,71 @@
+            PlaceCaret("Point");
+            VerifyCurrentTokenType("struct name");
+            RemoveMetadataReference("WindowsBase");
+            WaitForAsyncOperations(FeatureAttribute.Workspace);
```
No direct waiting in tests like this. Please move the wait to the appropriate helper.

jasonmalinowski commented on src/VisualStudio/IntegrationTest/IntegrationTests/Workspace/Workspaces.cs:
```diff
@@ -0,0 +1,71 @@
+            PlaceCaret("Class1");
+            VerifyCurrentTokenType("class name");
+            RemoveProjectReference(ProjectName, "CSProj2");
+            WaitForAsyncOperations(FeatureAttribute.Workspace);
```
No direct waiting in tests like this. Please move the wait to the appropriate helper.

jasonmalinowski commented on src/VisualStudio/IntegrationTest/IntegrationTests/Workspace/WorkspacesNetCore.cs:
```diff
@@ -0,0 +1,71 @@
+            EnableFullSolutionAnalysis();
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.Workspace)]
```
Aren't these the same tests as the other Workspaces tests? If so, please use xunit's support for inheritance to share the code.

jasonmalinowski commented on src/VisualStudio/IntegrationTest/IntegrationTests/Workspace/WorkspacesNetCore2.cs:
```diff
@@ -0,0 +1,42 @@
+namespace Roslyn.VisualStudio.IntegrationTests.Workspace
+{
+    [Collection(nameof(SharedIntegrationHostFixture))]
+    public class WorkspacesNetCore2 : AbstractEditorTest
```
Why 2?

jasonmalinowski commented on src/VisualStudio/IntegrationTest/TestUtilities/InProcess/SolutionExplorer_InProc.cs:
```diff
@@ -59,6 +59,23 @@ public static SolutionExplorer_InProc Create()
 
+        public void AddMetadataReference(string assemblyName, string projectName)
+        {
+            var referenceAssemblyFullPath = System.Reflection.Assembly.Load(assemblyName).Location;
```
VSProject lets you give assembly qualified names to it. Why can't we use that support? This is actually sketchy because we're doing a load in a different context (unit tests) which might cause strange things to happen.

jasonmalinowski commented on src/VisualStudio/IntegrationTest/TestUtilities/InProcess/SolutionExplorer_InProc.cs:
```diff
@@ -59,6 +59,23 @@ public static SolutionExplorer_InProc Create()
+        public void RemoveMetadataReference(string assemblyName, string projectName)
+        {
+            var project = GetProject(projectName);
+            if (project.Object is VSProject vsproject)
```
Please throw if the project isn't valid, since that means you're passing the wrong name.

jasonmalinowski commented on src/VisualStudio/IntegrationTest/IntegrationTests/VisualBasic/BasicWinForms.cs:
```diff
@@ -96,18 +96,18 @@ public void Change_Control_Property_In_Code()
+            OpenFile("Form1.vb", ProjectName);
             VerifyTextContains(@"Private Sub ExecuteWhenButtonClicked(sender As Object, e As EventArgs) Handles SomeButton.Click");
-            SaveFile(ProjectName, "Form1.vb");
+            SaveFile("Form1.vb", ProjectName);
```
Should we make a little "Project" type of some kind to wrap this so we have better type safety?

jasonmalinowski commented on src/VisualStudio/IntegrationTest/TestUtilities/InProcess/VisualStudioWorkspace_InProc.cs:
```diff
@@ -46,6 +55,27 @@ public void SetPrettyListing(string languageName, bool value)
+                throw new Exception($"Failed to find option with feature name '{feature}' and option name '{optionName}'");
+            }
+            var result = TypeDescriptor.GetConverter(option.Type).ConvertFromString(value);
+            var optionKey = string.IsNullOrWhiteSpace(language)
```
Why is the "SetPerLanguageOption" method allowing language to be optional?

jasonmalinowski commented on src/VisualStudio/IntegrationTest/TestUtilities/OutOfProcess/VisualStudioWorkspace_OutOfProc.cs:
```diff
@@ -20,12 +21,21 @@ public bool IsUseSuggestionModeOn()
+            => _inProc.SetOptionInfer(value);
+
+        public void SetPersistenceOption(bool value)
+            => SetPerLanguageOption("Enabled", "FeatureManager/Persistence", null, value ? "true" : "false");
```
Don't we already have internals visible to Roslyn to use the IOptionService? Can we just reference the option directly?

jasonmalinowski replied:
Why the conversion to true/false if you already have the TypeConverter code?

jasonmalinowski commented on src/VisualStudio/IntegrationTest/TestUtilities/OutOfProcess/SolutionExplorer_OutOfProc.cs:
```diff
@@ -37,6 +37,15 @@ public void AddProject(string projectName, string projectTemplate, string langua
         public void AddProjectReference(string fromProjectName, string toProjectName)
             => _inProc.AddProjectReference(fromProjectName, toProjectName);
 
+        public void RemoveProjectReference(string projectReferenceName, string projectName)
```
Aren't the arguments here swapped from AddProjectReference?

jasonmalinowski commented on src/VisualStudio/IntegrationTest/TestUtilities/InProcess/VisualStudioWorkspace_InProc.cs:
```diff
@@ -23,6 +25,13 @@ private VisualStudioWorkspace_InProc()
+        public void SetOptionInfer(bool value)
+            => InvokeOnUIThread(() => {
+                var convertedValue = value ? 1 : 0;
+                var project = GetDTE().Solution.Projects.Item(1);
```
Why doesn't this let you configure which project it is?

jasonmalinowski commented on src/VisualStudio/IntegrationTest/TestUtilities/InProcess/VisualStudioWorkspace_InProc.cs:
```diff
@@ -23,6 +25,13 @@ private VisualStudioWorkspace_InProc()
+                project.Properties.Item("OptionInfer").Value = convertedValue;
+            });
+
         public static VisualStudioWorkspace_InProc Create()
```
This should probably be up by the constructor.

jasonmalinowski commented on src/VisualStudio/IntegrationTest/IntegrationTests/Workspace/WorkspacesNetCore2.cs:
```diff
@@ -0,0 +1,42 @@
+            EnableFullSolutionAnalysis();
+        }
+
+        [Fact(Skip = "VB Not Supported"), Trait(Traits.Feature, Traits.Features.Workspace)]
```
So this has one test that's skipped...? :smile: Please revise to the tracking bug.

jasonmalinowski commented on src/VisualStudio/IntegrationTest/IntegrationTests/AbstractEditorTest.cs:
```diff
@@ -138,10 +159,52 @@ protected void SendKeys(params object[] keys)
+            WaitForAsyncOperations(FeatureAttribute.Workspace);
+        }
+
+        protected void EnableQuickInfo()
```
Should we just have the methods being Set* that take a Boolean rather than doubling all the methods here? This class's concept count is getting disturbingly big, disturbingly quick.

jmarolf replied:
We can.  I don't see how every test ever isn't going to derive from `AbstractEditorTest` at this point, there is too much useful functionality at this point.  Integration tests are naturally going to involve a broad set of functionality.  Sounds like out rules for tests are
- Move all helper methods into `AbstractEditorTest` 
- Use Assert in the test unless it is more than one line, then move that verification to a helper method in `AbstractEditorTest`

tmeschter replied:
No. Do not move all helper methods into `AbstractEditorTest`. I'm not a fan of the "dumping ground" design pattern. :-(

We shouldn't be using inheritance as a means of supplying _access_ to useful functionality. Code that is broadly useful should be grouped into classes of related functionality, like we've done with `SolutionExplorer_In\OutOfProc` for manipulating solutions and projects, the dialog types for manipulating specific dialogs, etc. Inheritance should be used to _customize_ behavior, like when we need each test class to supply the appropriate language name.

This strongly suggests that we shouldn't have an `AbstractEditorTest` class at all: most functionality should be moved out to helper types, or perhaps down to `AbstractIntegrationTest` itself.

jasonmalinowski commented on src/VisualStudio/IntegrationTest/IntegrationTests/AbstractEditorTest.cs:
```diff
@@ -138,10 +159,52 @@ protected void SendKeys(params object[] keys)
+
+        protected void DisableQuickInfo()
+        {
+            VisualStudioWorkspaceOutOfProc.EnableQuickInfo(false);
```
Not your doing, but can we rename this to SetQuickInfo or something?

tmeschter commented on src/VisualStudio/IntegrationTest/IntegrationTests/AbstractEditorTest.cs:
```diff
@@ -138,10 +159,52 @@ protected void SendKeys(params object[] keys)
+        }
+
+        protected void EnableOptionInfer()
+        {
```
For example, these Enabled*/Disable* helpers should be moved elsewhere.

CyrusNajmabadi commented on src/VisualStudio/IntegrationTest/IntegrationTests/Workspace/WorkspacesNetCore.cs:
```diff
@@ -31,15 +33,16 @@ public void OpenCSharpThenVBSolution()
-            SaveAll();
-            WaitForAsyncOperations(FeatureAttribute.Workspace);
-            OpenFile("Class1.cs");
+            this.SaveAll();
```
i believe the Roslyn style is to not have "this" on methods.

jmarolf replied:
It needs to have `this` because it is an extension method

jmarolf commented:
@tmeschter @jasonmalinowski @dpoeschl 
I've done a few things here.  

1. I've moved all of the functionality that was dumped into AbstractEditorTest into extension methods.  Here is my reasoning: integration tests are always going to contain a large cross section of functionality.  Thats the whole point.  While functional tests can be factored to only contain the functionality associated with their features under test, integration tests need everything. Using inheritance this would mean that one class would be the dumping ground for all the common functionality that everyone needs.  Instead I propose that all of this should go into extension methods, that way the test writer can import the functionality they need.  These extension methods are off the AbstractIntegrationTest type because I have heard feedback that people wanted a terse syntax for calling these and `this.Method()` is the best I could come up with.  Also, I only ever expect these methods to be called from within tests.

2. I've removed one line Verify methods and have just moved that functionality into the tests itself

3. I've added some wrapper types so methods that used to accept multiple string types now have some measure of type safety

jmarolf commented:
retest windows_debug_vs-integration_prtest please

jmarolf commented:
retest windows_release_vs-integration_prtest please

jmarolf commented:
retest windows_debug_vs-integration_prtest please

jmarolf commented:
retest windows_debug_vs-integration_prtest please

jmarolf commented:
retest windows_release_vs-integration_prtest please

jmarolf commented:
retest windows_debug_vs-integration_prtest please

jmarolf commented:
retest windows_release_vs-integration_prtest please

jmarolf commented:
retest please

jmarolf commented:
retest all please

jmarolf commented:
retest windows_release_vs-integration_prtest please
